File: 0013_roman_to_integer.go
package main

func romanToInt(s string) int {
	m := make(map[byte]int)
	m['I'] = 1
	m['V'] = 5
	m['X'] = 10
	m['L'] = 50
	m['C'] = 100
	m['D'] = 500
	m['M'] = 1000

	i := len(s) - 1
	sum := m[s[i]]
	i--
	for ; i >= 0; i-- {
		if m[s[i]] < m[s[i+1]] {
			sum -= m[s[i]]
		} else {
			sum += m[s[i]]
		}

	}
	return sum
}

File: 0013/0013_roman_to_integer.rs
struct Solution;

impl Solution {
    pub fn get_value(ch: char) -> i32 {
        match ch {
            'I' => 1,
            'V' => 5,
            'X' => 10,
            'L' => 50,
            'C' => 100,
            'D' => 500,
            'M' => 1000,
            _ => 0,
        }
    }

    pub fn roman_to_int(s: String) -> i32 {
        let mut sum = 0;
        let mut last = 0;
        for c in s.chars().rev() {
            let curr = Self::get_value(c);
            if curr >= last {
                sum += curr;
            } else {
                sum -= curr;
            }
            last = curr;
        }
        sum
    }
}

fn main() {
    assert_eq!(Solution::roman_to_int("III".to_string()), 3);
    assert_eq!(Solution::roman_to_int("IV".to_string()), 4);
}
